art/defences/trainer/dp_instahide_trainer.py

```diff
@@ -0,0 +1,285 @@
+                    train_loss / train_n,
+                    train_acc / train_n,
+                    test_loss,
+                    test_acc,
+                )
+            else:
+                logger.info(
+                    "epoch: %s time(s): %.1f, loss: %.4f, acc: %.4f",
+                    i_epoch,
+                    train_time - start_time,
+                    train_loss / train_n,
+                    train_acc / train_n,
+                )
+
+    def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwargs):
+        """
+        Train a model adversarially with the DP-InstaHide protocol using a data generator.
+        See class documentation for more information on the exact procedure.
+
+        :param generator: Data generator.
+        :param nb_epochs: Number of epochs to use for trainings.
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function
+               of the target classifier.
+        """
+        logger.info("Performing adversarial training with DP-InstaHide protocol")
+        size = generator.size
+        batch_size = generator.batch_size
+        if size is not None:
+            nb_batches = int(np.ceil(size / batch_size))
+        else:
+            raise ValueError("Size is None.")
+
+        logger.info("Adversarial Training DP-InstaHide")
+
+        for i_epoch in trange(nb_epochs, desc="DP-InstaHide training epochs"):
+            start_time = time.time()
+            train_loss = 0.0
+            train_acc = 0.0
+            train_n = 0.0
+
+            for _ in range(nb_batches):
+                # create batch data
+                x_batch, y_batch = generator.get_batch()
+
+                # augment the data
+                x_aug = x_batch.copy()
+                y_aug = y_batch.copy()
+                for augmentation in self.augmentations:
+                    x_aug, y_aug = augmentation(x_aug, y_aug)
+
+                # apply additive noise
+                x_aug = self._generate_noise(x_aug)
+
+                # extract label reduction and set to no reduction if needed
+                reduce_labels = self._classifier._reduce_labels  # type: ignore # pylint: disable=W0212
+                if len(y_aug.shape) == 2:
+                    self._classifier._reduce_labels = False  # type: ignore # pylint: disable=W0212
+
+                # fit batch
+                self._classifier.fit(x_aug, y_aug, nb_epochs=1, batch_size=x_aug.shape[0], verbose=0, **kwargs)
+
+                # get metrics
+                loss = self._classifier.compute_loss(x_aug, y_aug, reduction="mean")
+                output = np.argmax(self.predict(x_batch), axis=1)
+                acc = np.sum(output == np.argmax(y_batch, axis=1))
+                n = len(x_aug)
+
+                # restore original label reduction
+                self._classifier._reduce_labels = reduce_labels  # type: ignore # pylint: disable=W0212
```

Comment:
I think these lines should not be required anymore with https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/1960?

Reply:
Same as above